src/field_theory/separable_degree.lean

```diff
@@ -0,0 +1,81 @@
+import data.polynomial.degree
+import algebra.char_p.exp_char
+/-!
+# Separable degree
+
+This file contains basics about the separable degree of a polynomial.
+
+## Main results
+- `separable_degree`: the definition of the separable degree
+- `irreducible_has_separable_contraction`: any nonzero irreducible polynomial can be contracted
+  to a separable polynomial
+- `separable_degree_dvd_degree'`: the separable degree divides the degree, in function of the
+  exponential characteristic of the field
+- `separable_degree_dvd_degree` and `separable_degree_eq_degree` specialize the statement of
+  `separable_degree_dvd_degree`
+
+## Tags
+separable degree, degree, polynomial
+-/
+
+noncomputable theory
+open_locale classical
+
+variables {F : Type} [comm_semiring F]
+  (q : ℕ) [exp_char F q]
+
+/-- A separable contraction of a polynomial `f` is a separable polynomial `g` such that
+`g(x^(q^m)) = f(x)` for some `m : ℕ`.-/
+class has_separable_contraction (f : polynomial F) : Prop :=
+(has_separable_contraction' : ∃ g : polynomial F,
+  g.separable ∧ ∃ m : ℕ, polynomial.expand F (q^m) g = f)
+
+/-- A choice of a separable contraction. -/
+def separable_contraction (f : polynomial F) [hf : has_separable_contraction q f] :
+  polynomial F := classical.some hf.has_separable_contraction'
+
+/-- The separable degree of a polynomial is the degree of a given separable contraction. -/
+def separable_degree (f : polynomial F) [has_separable_contraction q f] : ℕ :=
+polynomial.nat_degree (separable_contraction q f)
+
+/-- Every irreducible polynomial can be contracted to a separable polynomial.
+https://stacks.math.columbia.edu/tag/09H0 -/
+lemma irreducible_has_separable_contraction {F : Type} [field F] (q : ℕ) [hF : exp_char F q]
+  (f : polynomial F) [irred : irreducible f] (fn : f ≠ 0) : has_separable_contraction q f :=
+begin
+  casesI hF,
+  { use f,
+    exact ⟨irreducible.separable irred, ⟨0, by rw [pow_zero, polynomial.expand_one]⟩⟩ },
+  { haveI qp : fact (nat.prime q) := fact_iff.mpr hF_hprime,
+    rcases polynomial.exists_separable_of_irreducible q irred fn with ⟨n, g, hg⟩,
+    use g,
+    exact ⟨hg.1, ⟨n, hg.2⟩⟩, }
+end
+
+/-- The separable degree divides the degree, in function of the exponential characteristic of F. -/
+lemma separable_degree_dvd_degree' (f : polynomial F) [hf : has_separable_contraction q f] :
+  ∃ m : ℕ, (separable_degree q f) * (q ^ m) = f.nat_degree :=
+begin
+  cases (classical.some_spec hf.has_separable_contraction').2 with m hm,
+  use m,
+  let g := classical.some hf.has_separable_contraction',
+  have deg_g : g.nat_degree = separable_degree q f, by refl,
+  rw [←deg_g, ←hm, polynomial.nat_degree_expand (q^m) g, mul_comm],
+end
+
+/-- The separable degree divides the degree. -/
+lemma separable_degree_dvd_degree (f : polynomial F) [hf : has_separable_contraction q f] :
+  (separable_degree q f) ∣ f.nat_degree :=
+exists.elim (separable_degree_dvd_degree' q f) (λ a, λ ha, dvd.intro (q ^ a) ha)
```

Comment:
```suggestion
exists.elim (separable_degree_dvd_degree' q f) (λ a, dvd.intro (q ^ a))
```
You don't need to intro `ha` if all you're going to do is feed it back into a function at the end :-)

Reply:
OK.